Distill trainer

orenpereg opened the pull request:
1. Added Trainer for setfit distillation (file src/setfit/trainer_distill.py)
2. updated fewshot_distillation script to work with DistilSetFitTrainer

lewtun commented on src/setfit/trainer_distill.py:
```diff
@@ -0,0 +1,239 @@
+
+        x_train = train_dataset["text"]
+        y_train = train_dataset["label"]
+        if self.loss_class is None:
```
Should we raise a proper `ValueError` here with a message or did I misunderstand the role of this `return` statement?

orenpereg replied:
i agree, i added a ValueError with a message. BTW, we had the same issue in SetFitTrainer i also added a ValueError there.

lewtun reviewed (commented):
Thanks for adding this nice feature @orenpereg 🔥 !

I've left a few comments / questions and would love to see some unit tests in a `test_trainer_distillation.py` file (you can take inspiration from the existing `SetFitTrainer` ones)

Could you also share a code snippet of how this trainer works end-to-end? I know we have that logic in the script, but it would be great to have a condensed version we can later add to the README :)

lewtun commented on scripts/setfit/run_fewshot_distillation.py:
```diff
@@ -245,18 +236,60 @@ def train(self):
+                        model=teacher_model,
+                        train_dataset=fewshot_ds[name],
+                        eval_dataset=eval_dataset,
+                        loss_class=CosineSimilarityLoss,
```
Should we add the type with something like this?

```suggestion
                        loss_class: torch.nn.Module =CosineSimilarityLoss,
```

orenpereg replied:
good point. done

orenpereg commented on scripts/setfit/run_fewshot_distillation.py:
```diff
@@ -10,22 +8,17 @@
 from distillation_baseline import BaselineDistillation
 from evaluate import load
+from sentence_transformers import losses
+from sentence_transformers.losses import CosineSimilarityLoss
```
thanks, done.

orenpereg commented on scripts/setfit/run_fewshot_distillation.py:
```diff
@@ -84,7 +77,7 @@ def parse_args():
 
 
 class RunFewShotDistill:
```
I ran the ag_news test with using 5 seeds. the results are very close. the diffs are between 0.1-0.6 points. i also checked emotion using a single seed and it looks quite similar. BTW the baseline model produces lower results. i still have to check why.
![Screen Shot 2022-11-13 at 18 01 20](https://user-images.githubusercontent.com/31614579/201531472-128927b9-2e94-434f-94ce-d5e07eff32f6.png)

orenpereg commented on scripts/setfit/run_fewshot_distillation.py:
```diff
@@ -245,18 +236,60 @@ def train(self):
                 # if os.path.exists(results_path):
```
yes, done

orenpereg commented on scripts/setfit/run_fewshot_distillation.py:
```diff
@@ -245,18 +236,60 @@ def train(self):
-                        self.args.num_epochs,
-                        metric,
+                if self.mode == self.TEACHER:
+                    # self.model.load_state_dict(copy.deepcopy(self.model_wrapper.model_original_state))
```
yes, done.

orenpereg commented on scripts/setfit/run_fewshot_distillation.py:
```diff
@@ -271,83 +304,17 @@ def train_baseline_student(self, train_data, test_data, num_classes):
 
-        x_train = self.x_train_teacher + x_train
-        x_train_embd_student = self.trained_teacher_model.sbert_model.encode(x_train)
+        # x_train_embd_student = self.trained_teacher_model.sbert_model.encode(x_train)
```
yes, done :)

orenpereg commented on src/setfit/trainer_distill.py:
```diff
@@ -0,0 +1,239 @@
+from torch.utils.data import DataLoader
+from transformers.trainer_utils import set_seed
+
+from setfit import SetFitTrainer
```
thanks! good point, done.

orenpereg commented on src/setfit/trainer_distill.py:
```diff
@@ -0,0 +1,239 @@
+
+
+class DistilSetFitTrainer(SetFitTrainer):
+    """Trainer to train a distilled SetFit model (student) model.
```
yes, this sounds much better. done.

orenpereg commented on src/setfit/trainer_distill.py:
```diff
@@ -0,0 +1,239 @@
+
+    def __init__(
+        self,
+        teacher_model: "SetFitModel" = None,
```
Right, thanks, i removed 'optional'

orenpereg commented on scripts/setfit/run_fewshot_distillation.py:
```diff
@@ -245,18 +236,60 @@ def train(self):
+                        loss_class=CosineSimilarityLoss,
+                        metric=metric,
+                        batch_size=self.args.batch_size,
+                        num_iterations=20,  # The number of text pairs to generate for contrastive learning
```
I agree. i changed to 'args.num_iterations_student' and 'args.num_iterations_teacher'.

orenpereg commented on scripts/setfit/run_fewshot_distillation.py:
```diff
@@ -245,18 +236,60 @@ def train(self):
+                        loss_class=CosineSimilarityLoss,
+                        metric="accuracy",
+                        batch_size=16,
+                        num_iterations=20,  # The number of text pairs to generate for contrastive learning
```
yes. done :)

orenpereg commented on scripts/setfit/run_fewshot_distillation.py:
```diff
@@ -245,18 +236,60 @@ def train(self):
+                    # student train data = teacher train data + unlabeled data
+                    student_train_dataset = concatenate_datasets([self.treacher_train_data, fewshot_ds[name]])
+                    student_train_dataset.rename_column(
+                        "text", "sentence"
```
You are right, I deleted it. the columns are normalized under the setfit org.

orenpereg commented:
Hi @lewtun. Thanks for your thorough review and good comments🔥. I addressed all your comments and added tests under tests/test_trainer_distillation.py. can you pls take a look?
I also wrote a code snippet for the readme for using this trainer end-to-end:
[Run distillation using trainer example.docx](https://github.com/huggingface/setfit/files/10010292/Run.distillation.using.trainer.example.docx)

lewtun commented on scripts/setfit/run_fewshot_distillation.py:
```diff
@@ -245,18 +236,60 @@ def train(self):
+                        batch_size=16,
+                        num_iterations=20,  # The number of text pairs to generate for contrastive learning
+                        num_epochs=1  # The number of epochs to use for constrastive learning
+                        # column_mapping={"sentence": "text", "label": "label"} # Map dataset columns to text/label expected by trainer
```
I think the unused code is still here :)

lewtun reviewed (commented):
Thanks for iterating on this feature @orenpereg and adding some nice tests ❤️ !

I've left some final nits and then this should be good to merge :)

orenpereg commented on scripts/setfit/run_fewshot_distillation.py:
```diff
@@ -10,28 +8,21 @@
-)
+
+from setfit import DistillationSetFitTrainer, SetFitModel, SetFitTrainer
+from setfit.modeling import SetFitBaseModel, SKLearnWrapper
```
Thanks Lewis, I deleted it and also deleted the get_classifier function (that used SKLearnWrapper) which is not used

orenpereg commented on scripts/setfit/run_fewshot_distillation.py:
```diff
@@ -84,7 +78,7 @@ def parse_args():
 
 class RunFewShotDistill:
-    def __init__(self, args, mode, trained_teacher_model, x_train_teacher, student_train_ds) -> None:
+    def __init__(self, args, mode, trained_teacher_model, treacher_train_data, student_train_ds) -> None:
```
Thanks, good idea. done.

orenpereg commented on src/setfit/trainer.py:
```diff
@@ -287,7 +287,7 @@ def train(
         y_train = train_dataset["label"]
         if self.loss_class is None:
-            return
+            raise RuntimeError("SetFitTrainer: training requires a loss_class argument")
```
very good idea. this is much cleaner. i updated it in both Trainers

orenpereg commented on src/setfit/trainer_distillation.py:
```diff
@@ -0,0 +1,240 @@
+        eval_dataset (`Dataset`, *optional*):
+            The evaluation dataset.
+        model_init (`Callable[[], SetFitModel]`, *optional*):
+            A function that instantiates the model to be used. If provided, each call to [`~DistilledSetFitTrainer:.train`] will start
```
done

orenpereg commented on src/setfit/trainer_distillation.py:
```diff
@@ -0,0 +1,240 @@
+
+    def __init__(
+        self,
+        teacher_model: "SetFitModel" = None,
```
right. done

orenpereg commented on src/setfit/trainer_distillation.py:
```diff
@@ -0,0 +1,240 @@
+            set_seed(self.seed)  # Seed must be set before instantiating the model when using model_init.
+            self._hp_search_setup(trial)  # sets trainer parameters and initializes model
+
+        if self.teacher_model is None:
```
yes. done.

orenpereg commented on src/setfit/trainer_distillation.py:
```diff
@@ -0,0 +1,240 @@
+        if self.teacher_model is None:
+            raise RuntimeError("`DistillationSetFitTrainer` training requires a teacher model")
+
+        if self.train_dataset is None:
```
done

orenpereg commented on tests/test_trainer_distillation.py:
```diff
@@ -0,0 +1,112 @@
+        formatted_dataset = trainer._apply_column_mapping(trainer.train_dataset, trainer.column_mapping)
+
+        assert formatted_dataset.column_names == ["text", "label"]
+
```
done

orenpereg commented on tests/test_trainer_distillation.py:
```diff
@@ -0,0 +1,112 @@
+
+        assert formatted_dataset[0]["text"] == "a"
+        assert formatted_dataset[0]["label"] == [0, 1]
+
```
done

orenpereg commented on src/setfit/trainer_distillation.py:
```diff
@@ -0,0 +1,240 @@
+            The teacher model to mimic.
+        model (`SetFitModel`):
+            The student model to train. If not provided, a `model_init` must be passed.
+        train_dataset (`Dataset`):
```
thanks! i changes the order of the args and renamed model to student_model

lewtun reviewed (approved):
Thanks for the last round of changes @orenpereg - this looks really nice 🔥

orenpereg commented:
thanks Lewis :)
